src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ImmutableArrayExtensions.cs

```diff
@@ -16,11 +18,69 @@ public static ImmutableArray<T> NullToEmpty<T>(this ImmutableArray<T> array)
         return array.IsDefault ? ImmutableArray<T>.Empty : array;
     }
 
-    public static void SetCapacityIfNeeded<T>(this ImmutableArray<T>.Builder builder, int newCapacity)
+    public static void SetCapacityIfLarger<T>(this ImmutableArray<T>.Builder builder, int newCapacity)
     {
         if (builder.Capacity < newCapacity)
         {
             builder.Capacity = newCapacity;
         }
     }
+
+    /// <summary>
+    ///  Returns the current contents as an <see cref="ImmutableArray{T}"/> and sets
+    ///  the collection to a zero length array.
+    /// </summary>
+    /// <remarks>
+    ///  If <see cref="ImmutableArray{T}.Builder.Capacity"/> equals
+    ///  <see cref="ImmutableArray{T}.Builder.Count"/>, the internal array will be extracted
+    ///  as an <see cref="ImmutableArray{T}"/> without copying the contents. Otherwise, the
+    ///  contents will be copied into a new array. The collection will then be set to a
+    ///  zero-length array.
+    /// </remarks>
+    /// <returns>An immutable array.</returns>
+    public static ImmutableArray<T> DrainToImmutable<T>(this ImmutableArray<T>.Builder builder)
+    {
+#if NET8_0_OR_GREATER
+        return builder.DrainToImmutable();
+#else
+        if (builder.Count == 0)
+        {
+            return ImmutableArray<T>.Empty;
+        }
+
+        if (builder.Count == builder.Capacity)
+        {
+            return builder.MoveToImmutable();
+        }
+
+        var result = builder.ToImmutable();
+        builder.Clear();
+        return result;
+#endif
+    }
+
+    public static ImmutableArray<TResult> SelectAsArray<T, TResult>(this ImmutableArray<T> source, Func<T, TResult> selector)
+    {
+        return source switch
+        {
+            [] => ImmutableArray<TResult>.Empty,
+            [var item] => ImmutableArray.Create(selector(item)),
+            [var item1, var item2] => ImmutableArray.Create(selector(item1), selector(item2)),
+            [var item1, var item2, var item3] => ImmutableArray.Create(selector(item1), selector(item2), selector(item3)),
+            [var item1, var item2, var item3, var item4] => ImmutableArray.Create(selector(item1), selector(item2), selector(item3), selector(item4)),
+            var items => BuildResult(items, selector)
+        };
+
+        static ImmutableArray<TResult> BuildResult(ImmutableArray<T> items, Func<T, TResult> selector)
+        {
+            using var _ = ArrayBuilderPool<TResult>.GetPooledObject(out var result);
```

Comment:
Consider ensuring the builder has capacity for all `items` to avoid resizes.

Reply:
Smiling because I actually took the time to rename the extension method to `SetCapacityIfLarger` and then forgot to use it! :smile: